module/modmanager/manager.go

```diff
@@ -716,29 +730,27 @@ func (mgr *Manager) attemptRestart(ctx context.Context, mod *module) []resource.
 	return nil
 }
 
-// dial will use m.conn to make a new module service client or Dial m.addr if
-// m.conn is nil.
+// dial will Dial the module and replace the underlying connection (if it exists) in m.conn.
 func (m *module) dial() error {
-	if m.conn == nil {
-		// TODO(PRODUCT-343): session support probably means interceptors here
-		var err error
-		m.conn, err = grpc.Dial(
-			"unix://"+m.addr,
-			grpc.WithTransportCredentials(insecure.NewCredentials()),
-			grpc.WithChainUnaryInterceptor(
-				grpc_retry.UnaryClientInterceptor(),
-				operation.UnaryClientInterceptor,
-			),
-			grpc.WithChainStreamInterceptor(
-				grpc_retry.StreamClientInterceptor(),
-				operation.StreamClientInterceptor,
-			),
-		)
-		if err != nil {
-			return errors.WithMessage(err, "module startup failed")
-		}
+	// TODO(PRODUCT-343): session support probably means interceptors here
+	var err error
+	conn, err := grpc.Dial(
+		"unix://"+m.addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithChainUnaryInterceptor(
+			grpc_retry.UnaryClientInterceptor(),
+			operation.UnaryClientInterceptor,
+		),
+		grpc.WithChainStreamInterceptor(
+			grpc_retry.StreamClientInterceptor(),
+			operation.StreamClientInterceptor,
+		),
+	)
+	if err != nil {
+		return errors.WithMessage(err, "module startup failed")
 	}
-	m.client = pb.NewModuleServiceClient(m.conn)
+	m.conn.ReplaceConn(conn)
```

Comment:
should we also close the old connection?

Reply:
the old connection gets closed during `closeModule` and `newOnUnexpectedExitHandler` - I don't think we need to close it again

Reply:
I'm almost certainly missing something obvious, but in both of cases we close the current connection - it's not clear to me that we close the connections that may have been swapped out.

Reply:
okay I tracked it down... it does seem like do always close the older connection before we replace it. I wonder if this should just happen within replace?

definitely not a change to make in this PR though since this behavior existed before this change